module/zfs/dsl_dataset.c

```diff
@@ -1273,9 +1273,9 @@ dsl_dataset_zero_zil(dsl_dataset_t *ds, dmu_tx_t *tx)
 		dsl_dataset_sync(ds, zio, tx);
 		VERIFY0(zio_wait(zio));
 
-		/* dsl_dataset_sync_done will drop this reference. */
 		dmu_buf_add_ref(ds->ds_dbuf, ds);
 		dsl_dataset_sync_done(ds, tx);
+		dmu_buf_rele(ds->ds_dbuf, ds);
```

Comment:
We shouldn't need to add and then remove the reference.  the dataset remains valid across the call to dsl_dataset_sync_done().

Reply:
Yeah, I think this was originally structured this way because the rele was being done inside of the sync_done call.

Reply:
Fixed.